sw/device/sca/aes_serial.c

```diff
@@ -178,6 +195,96 @@ static void aes_serial_batch_encrypt(const uint8_t *data, size_t data_len) {
+ * attacks. Instead of expecting a plaintext and sending the resulting
+ * ciphertext from and to the host for each encryption, this command repeatedly
+ * encrypts random plaintexts that are generated on the device. The data
+ * collection method is based on the derived test requirements (DTR) for TVLA:
+ * https://www.rambus.com/wp-content/uploads/2015/08/TVLA-DTR-with-AES.pdf
+ * The measurements are taken by using either fixed or randomly selected keys.
+ * In order to simplify the analysis, the first encryption has to use fixed key.
+ * In addition a PRNG is used for random key and plaintext generation instead
+ * of AES algorithm as specified in the TVLA DTR.
+ * This minimizes the overhead of UART communication and significantly improves
+ * the capture rate. The host must use the same PRNG to be able to compute the
+ * random plaintext, random key and the ciphertext of each trace.
+ *
+ * Packet payload must be a `uint32_t` representation of the number of
+ * encryptions to perform. Since generated plaintexts are not cached, there is
+ * no limit on the number of encryptions.
+ *
+ * The PRNG should be initialized using the 's' (seed PRNG) command before
+ * starting batch encryption. In addition, the fixed key should also be set
+ * using 't' (fvsr key set) command before starting batch encryption.
+ *
+ * Note that the host can partially verify this operation by checking the
+ * contents of the 'r' (ciphertext) packet that is sent at the end.
+ *
+ * @param data Packet payload.
+ * @param data_len Packet payload length.
+ */
+static void aes_serial_fvsr_key_batch_encrypt(const uint8_t *data,
+                                              size_t data_len) {
+  uint32_t num_encryptions = 0;
+  SS_CHECK(data_len == sizeof(num_encryptions));
+  num_encryptions = read_32(data);
+
+  uint8_t key_random[kAesKeyLength];
+  uint8_t plaintext[kAesTextLength];
+  dif_aes_data_t ciphertext;
+  uint8_t ciphertext_total[kAesTextLength] = {0};
+
+  for (uint32_t i = 0; i < num_encryptions; ++i) {
+    // set key, plaintext and then start encryption
+    if (sample_fixed && !sample_fixed_prev) {
+      aes_serial_set_key(key_fixed, kAesKeyLength);
+    } else if (!sample_fixed) {
+      prng_rand_bytes(key_random, kAesKeyLength);
+      aes_serial_set_key(key_random, kAesKeyLength);
+    }
+    prng_rand_bytes(plaintext, kAesTextLength);
+    sca_set_trigger_high();
+    aes_serial_encrypt(plaintext, kAesTextLength);
+    sca_set_trigger_low();
+
+    sample_fixed_prev = sample_fixed;
+    sample_fixed = plaintext[0] % 2;
+
+    // wait until AES output is valid and then xor it with the previous
+    // ciphertexts in the batch to check the result at the end of every batch
+    bool ready = false;
+    do {
+      SS_CHECK_DIF_OK(
+          dif_aes_get_status(&aes, kDifAesStatusOutputValid, &ready));
+    } while (!ready);
+    SS_CHECK_DIF_OK(dif_aes_read_output(&aes, &ciphertext));
```

Comment:
Maybe it would be worth to just read and XOR the last data word to improve performance.

Reply:
It only has increased performance from 320 traces/s to 329 traces/s.